eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Imp/EnactSpec.hs

```diff
@@ -79,145 +100,343 @@ spec =
+          `shouldReturn` SJust (GovPurposeId gai2)
+        checkProposalsEmpty
+
+        committee <-
+          getsNES $
+            nesEsL . esLStateL . lsUTxOStateL . utxosGovStateL . committeeGovStateL
+        committee `shouldBe` SNothing
+
+    let val1 = Coin 1_000_001
+    let val2 = Coin 1_000_002
+    let val3 = Coin 1_000_003
 
-      impAnn "Constitution is enacted after two epochs" $ do
-        curConstitution <- getsNES $ newEpochStateGovStateL . constitutionGovStateL
-        curConstitution `shouldBe` constitution
+    it "The first submitted proposal of the same priority gets enacted first" $ do
+      (drepC, committeeC, _) <- electBasicCommittee
+      modifyPParams $ ppGovActionLifetimeL .~ EpochInterval 5
+      pGai0 <-
+        submitParameterChange
+          SNothing
+          $ def & ppuDRepDepositL .~ SJust val1
+      pGai1 <-
+        submitParameterChange
+          (SJust $ GovPurposeId pGai0)
+          $ def & ppuDRepDepositL .~ SJust val2
+      pGai2 <-
+        submitParameterChange
+          (SJust $ GovPurposeId pGai1)
+          $ def & ppuDRepDepositL .~ SJust val3
+      traverse_ @[]
+        ( \gaid -> do
+            submitYesVote_ (DRepVoter drepC) gaid
+            submitYesVote_ (CommitteeVoter committeeC) gaid
+        )
+        [pGai0, pGai1, pGai2]
+      passNEpochs 2
+      getLastEnactedParameterChange
+        `shouldReturn` SJust (GovPurposeId pGai2)
+      checkProposalsEmpty
+      getsNES (nesEsL . curPParamsEpochStateL . ppDRepDepositL)
+        `shouldReturn` val3
+
+    it "If submitted in the same transaction, only the first action gets enacted" $ do
+      (drepC, committeeC, _) <- electBasicCommittee
+      modifyPParams $ ppGovActionLifetimeL .~ EpochInterval 5
+      gaids <-
+        submitGovActions $
+          NE.fromList
+            [ ParameterChange
+                SNothing
+                (def & ppuDRepDepositL .~ SJust val1)
+                SNothing
+            , ParameterChange
+                SNothing
+                (def & ppuDRepDepositL .~ SJust val2)
+                SNothing
+            , ParameterChange
+                SNothing
+                (def & ppuDRepDepositL .~ SJust val3)
+                SNothing
+            ]
+      traverse_
+        ( \gaid -> do
+            submitYesVote_ (DRepVoter drepC) gaid
+            submitYesVote_ (CommitteeVoter committeeC) gaid
+        )
+        gaids
+      passNEpochs 2
+      getsNES (nesEsL . curPParamsEpochStateL . ppDRepDepositL)
+        `shouldReturn` val1
+      checkProposalsEmpty
 
-      impAnn "Pulser is reset" $ do
-        ConwayGovState _ _ _ _ _ pulser <- getsNES newEpochStateGovStateL
-        let pulserRatifyState = extractDRepPulsingState pulser
-        rsEnacted pulserRatifyState `shouldBe` Seq.empty
-        enactState <- getEnactState
-        rsEnactState pulserRatifyState `shouldBe` enactState
+checkProposalsEmpty :: ConwayEraGov era => ImpTestM era ()
```

Comment:
Let's move this to `Test.Cardano.Ledger.Conway.ImpTest`

Reply:
Also I think we should case match on `Proposals` and then give a more informative error message, right now it says `Expected True, got False` or something like that

Reply:
You can copy from [my PR](https://github.com/IntersectMBO/cardano-ledger/pull/4205/files#diff-832fb7c0b3f3c6b053b227f99393da3cd4ae7bea71729df757f8b8823139021b)

Reply:
Nice, thanks both, I replaced with `expectNoCurrentProposals`